source/h2_frames.c

```diff
@@ -332,714 +256,829 @@ void aws_h2_frame_encoder_clean_up(struct aws_h2_frame_encoder *encoder) {
+        goto handle_waiting_for_more_space;
+    }
 
-    return aws_h2_frame_header_block_init(&frame->header_block, allocator);
-}
-void aws_h2_frame_push_promise_clean_up(struct aws_h2_frame_push_promise *frame) {
-    AWS_PRECONDITION(frame);
+    const size_t fragment_len = aws_min_size(max_fragment, frame->header_block_cursor.len);
+    if (fragment_len == frame->header_block_cursor.len) {
+        /* This will finish the header-block */
+        flags |= AWS_H2_FRAME_F_END_HEADERS;
+    } else {
+        /* If we're not finishing the header-block, is it even worth trying to send this frame now? */
+        const size_t even_worth_sending_threshold = s_frame_prefix_length + payload_overhead;
+        if (fragment_len < even_worth_sending_threshold) {
+            goto handle_waiting_for_more_space;
+        }
+    }
 
-    aws_h2_frame_header_block_clean_up(&frame->header_block);
-}
+    /*
+     * Ok, it fits! Write the frame
+     */
+    bool all_wrote = true;
 
-int aws_h2_frame_push_promise_encode(
-    struct aws_h2_frame_push_promise *frame,
-    struct aws_h2_frame_encoder *encoder,
-    struct aws_byte_buf *output) {
-    AWS_PRECONDITION(frame);
-    AWS_PRECONDITION(encoder);
-    AWS_PRECONDITION(output);
+    /* Write the frame prefix */
+    const size_t payload_len = fragment_len + payload_overhead;
+    s_frame_prefix_encode(frame_type, frame->base.stream_id, payload_len, flags, output);
 
-    if (frame->promised_stream_id & s_u32_top_bit_mask) {
-        return aws_raise_error(AWS_ERROR_HTTP_PROTOCOL_ERROR);
+    /* Write pad length */
+    if (flags & AWS_H2_FRAME_F_PADDED) {
+        AWS_ASSERT(frame_type != AWS_H2_FRAME_T_CONTINUATION);
+        all_wrote &= aws_byte_buf_write_u8(output, pad_length);
     }
 
-    const size_t output_init_len = output->len;
-
-    /* Write header */
-    uint8_t flags = 0;
-    size_t length = 0;
-    if (aws_h2_frame_header_block_get_encoded_length(&frame->header_block, encoder, &length)) {
-        goto compression_error;
-    }
-    length += 4; /* Account for promised stream id */
-    if (frame->pad_length) {
-        flags |= AWS_H2_FRAME_F_PADDED;
-        length += frame->pad_length + 1;
-    }
-    if (frame->end_headers) {
-        flags |= AWS_H2_FRAME_F_END_HEADERS;
-    }
-    if (s_frame_prefix_encode(&frame->base, length, flags, output)) {
-        goto write_error;
+    /* Write priority */
+    if (flags & AWS_H2_FRAME_F_PRIORITY) {
+        AWS_ASSERT(frame_type == AWS_H2_FRAME_T_HEADERS);
+        s_frame_priority_settings_encode(priority_settings, output);
     }
 
-    /* Write pad length */
-    if (frame->pad_length) {
-        aws_byte_buf_write_u8(output, frame->pad_length);
+    /* Write promised stream ID */
+    if (promised_stream_id) {
+        AWS_ASSERT(frame_type == AWS_H2_FRAME_T_PUSH_PROMISE);
+        all_wrote &= aws_byte_buf_write_be32(output, *promised_stream_id & s_31_bit_mask);
```

Comment:
if you've already validated the stream_id do you actually need the mask?

Reply:
guess not anymore. will change